### PR TITLE
Introduce new CRD for EgressGW

### PR DESCRIPTION
**Pull request by julianwiedmann**
As part of promoting the EgressGW feature to stable, introduce a new CRD type and deprecate the old one.  The new CRD allows for better selection of the Egress Node & IP.

```release-note
Introduce a new CRD (CiliumEgressGatewayPolicy) for Egress Gateway configuration. Deprecate the previous CRD (CiliumEgressNATPolicy).
```

### CODE REVIEW

File: test/k8s/manifests/egress-nat-policy.yaml [hunk 1]
```diff
@@ -63,12 +63,12 @@ spec:
         cilium.io/ci-node: k8s2
     egressIP: INPUT_EGRESS_IP
 ---
-apiVersion: cilium.io/v2alpha1
-kind: CiliumEgressNATPolicy
+apiVersion: cilium.io/v2
```
Comment (jibi) [inline test/k8s/manifests/egress-nat-policy.yaml]: This should stay `v2alpha1` as it's still using the old `egressSourceIP` field (beside that tests seem happy).
We can then remove it once we get rid of the old API

File: pkg/egressgateway/policy.go [hunk 2]
```diff
@@ -299,3 +286,94 @@ func ParsePolicyConfigID(cenp *v2alpha1.CiliumEgressNATPolicy) types.NamespacedN
 		Name: cenp.Name,
 	}
 }
+
+// ParseCEGP takes a CiliumEgressGatewayPolicy CR and converts to PolicyConfig,
+// the internal representation of the egress gateway policy
+func ParseCEGP(cegp *v2.CiliumEgressGatewayPolicy) (*PolicyConfig, error) {
```
Comment (jibi) [inline pkg/egressgateway/policy.go:292]: nit: maybe rename also `ParsePolicy` to `ParseCENP`? (Or even better if possible refactor the 2 into a single function that takes the API version as parameter)

Comment (julianwiedmann) [reply pkg/egressgateway/policy.go:292]: Rename sounds good! Will have another look at consolidating the two functions, but maybe rather as a follow-up.

File: Documentation/contributing/testing/e2e.rst [hunk 3]
```diff
@@ -637,6 +637,8 @@ configuration options that can be passed as environment variables:
 +======================+===================+==============+==================================================================+
 | K8S\_NODES           | 2                 | 0..100       | Number of Kubernetes nodes in the cluster                        |
 +----------------------+-------------------+--------------+------------------------------------------------------------------+
+| NO_CILIUM_ON_NODE[S] | none              | *            | Comma-separated list of K8s nodes that should not run Cilium     |
```
Comment (qmonnet) [inline Documentation/contributing/testing/e2e.rst]: Nit: the star is rendered as a list item bullet, you might want to escape it with a backslash.
image

**Review by qmonnet — APPROVED** [review]

File: pkg/k8s/apis/cilium.io/v2/types.go [hunk 4]
```diff
@@ -13,9 +13,110 @@ import (
...
 )
 
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:categories={cilium,ciliumpolicy},singular="ciliumegressgatewaypolicy",path="ciliumegressgatewaypolicies",scope="Cluster",shortName={cegp}
+// +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
+// +kubebuilder:storageversion
+
+type CiliumEgressGatewayPolicy struct {
```
Comment (jibi) [inline pkg/k8s/apis/cilium.io/v2/types.go]: This should probably live in its own file (`cegp_types.go`)

File: Documentation/contributing/testing/e2e.rst (hunk 3 above)
Comment (julianwiedmann) [reply Documentation/contributing/testing/e2e.rst]: Ah indeed - thanks, fixed. Also fixed up the same issue for `SERVER_BOX`.

File: test/k8s/egress.go [hunk 5]
```diff
@@ -264,6 +264,37 @@ var _ = SkipDescribeIf(func() bool {
 					testConnectivity(true, ciliumOpts)
 				})
 			})
+
+			Context("egress gw policy upgrade", func() {
```
Comment (jibi) [inline test/k8s/egress.go:268]: Nice :100:

File: pkg/k8s/apis/cilium.io/v2alpha1/types.go [hunk 6]
```diff
@@ -14,9 +14,10 @@ import (
 // +genclient
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
-// +kubebuilder:resource:categories={cilium,ciliumpolicy},singular="ciliumegressnatpolicy",path="ciliumegressnatpolicies",scope="Cluster"
+// +kubebuilder:resource:categories={cilium,ciliumpolicy},singular="ciliumegressnatpolicy",path="ciliumegressnatpolicies",scope="Cluster",shortName={cenp}
 // +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
 // +kubebuilder:storageversion
+// +kubebuilder:deprecatedversion:warning="CiliumEgressNATPolicy is deprecated and will be removed in version 1.13. Use CiliumEgressGatewayPolicy instead."
```
Comment (jibi) [inline pkg/k8s/apis/cilium.io/v2alpha1/types.go:20]: TIL `kubebuilder:deprecatedversion`

**Review by jibi — APPROVED** [review]
LGTM :ship: :ship:

File: pkg/k8s/apis/cilium.io/v2/types.go (hunk 4 above)
Comment (julianwiedmann) [reply pkg/k8s/apis/cilium.io/v2/types.go]: Can do, was just sticking to the `v2alpha1` layout.

**Comment by julianwiedmann** [comment]
Resolved a conflict around `CustomResourceDefinitionSchemaVersion`, and moved things off into a separate `cegp_types.go` as suggested by @jibi .

**Comment by jibi** [comment]
@julianwiedmann you may want to rebase to pull in https://github.com/cilium/cilium/pull/19710 and fix all the K8sUpdates failures

**Comment by joestringer** [comment]
@julianwiedmann this PR is marked as a blocker, but I don't see any activity in the last 7 days. What's the next steps here?

**Comment by julianwiedmann** [comment]
> @julianwiedmann this PR is marked as a blocker, but I don't see any activity in the last 7 days. What's the next steps here?

Reviews essentially. There's nothing controversial in here I think.

[push to resolve a trivial merge conflict]

**Review by ldelossa — APPROVED** [review]
This all LGTM. Reading this review made it dawn on me that I did not add anything about watchers.go to the CRD documentation I submitted a bit earlier. If you'd like you can also update those user facing docs with instructions on how to plug into watcher's infrastructure.

**Review by nathanjsweet — APPROVED** [review]

**Review by nebril — APPROVED** [review]

**Pull request merged**